File: contest/lccup-2020-fall/3.qiu-xie-shou-cang-ji.py
# 3. 秋叶收藏集
# https://leetcode-cn.com/problems/UlBDOe/


class Solution1:
    def minimumOperations(self, leaves: str) -> int:
        n = len(leaves)
        f = [[0, 0, 0] for _ in range(n)]
        f[0][0] = int(leaves[0] == "y")
        f[0][1] = f[0][2] = f[1][2] = float("inf")

        for i in range(1, n):
            isRed = int(leaves[i] == "r")
            isYellow = int(leaves[i] == "y")
            f[i][0] = f[i - 1][0] + isYellow
            f[i][1] = min(f[i - 1][0], f[i - 1][1]) + isRed
            if i >= 2:
                f[i][2] = min(f[i - 1][1], f[i - 1][2]) + isYellow

        return f[n - 1][2]


class Solution:
    # TODO
    def minimumOperations(self, leaves: str) -> int:
        n = len(leaves)
        g = 1 if leaves[0] == "y" else -1

        gmin = g
        ans = float("inf")

        for i in range(1, n):
            isYellow = int(leaves[i] == "y")
            g += 2 * isYellow - 1
            if i != n - 1:
                ans = min(ans, gmin - g)
            gmin = min(gmin, g)

        return ans + (g + n) // 2


sol = Solution()
leaves = "rrryyyrryyyrr"
print(sol.minimumOperations(leaves))

File: contest/lccup-2020-fall/3.qiu-xie-shou-cang-ji.cpp
#include <string>
#include <iostream>

using namespace std;

class Solution
{
    int s[100005];

public:
    int minimumOperations(string leaves)
    {
        int n = leaves.size(), i, j, ans = n;
        for (i = 0; i < n; i++)
            s[i + 1] = s[i] + (leaves[i] == 'y');
        for (i = 1; i < n; i++)
            s[i] = i - 2 * s[i];
        for (i = 2, j = s[1]; i < n; i++)
        {
            ans = min(ans, s[n] + s[i] - j);
            j = max(j, s[i]);
        }
        return ans;
    }
};

int main()
{
    string s = "rrryyyrryyyrr";
    Solution sol;
    int res = sol.minimumOperations(s);
    cout << res;
}